File: Data Structures & Algorithms/decode-ways/submission-3.cpp
class Solution {
public:
    int numDecodings(string s) {
        int sz = s.size(), prev1 = 0, prev2 = 0, curr = 0;

        if (sz == 1) return s[0] != '0' ? 1 : 0;

        prev2 = s[sz - 1] != '0' ? 1 : 0;

        prev1 = s[sz - 2] != '0' ? prev2 : 0;
        prev1 += s[sz - 2] == '1' || (s[sz - 2] == '2' && s[sz - 1] <= '6') ? 1 : 0;

        for (int i = sz - 3; i >= 0; --i) {
            curr = 0;
            if (s[i] != '0')
                curr += prev1;

            if (s[i] == '1' || (s[i] == '2' && s[i + 1] <= '6'))
                curr += prev2;

            prev2 = prev1;
            prev1 = curr;
        }

        return prev1;
    }
};

File: Data Structures & Algorithms/course-schedule-ii/submission-3.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.adj = [[] for _ in range(numCourses)]
        for p in prerequisites:
            self.adj[p[0]].append(p[1])

        self.colors = [0] * numCourses

        self.ans = []

        for i in range(numCourses):
            if self.dfs(i):
                return []

        return self.ans
    def dfs(self, curr):
        if self.colors[curr] == 1:
            return True
        if self.colors[curr]:
            return False
        
        self.colors[curr] = 1

        for prereq in self.adj[curr]:
            if self.dfs(prereq): 
                return True
        
        self.colors[curr] = 2
        self.ans.append(curr)